### PR TITLE
[botcom] sharing

### PR DESCRIPTION
**Pull request by ds300**
This PR adds initial support for sharing.

- Using the file's `shared` flag to gate access to guests https://github.com/tldraw/tldraw/blob/1879c3a4b41529bb6b42992410ec5cf0ba3b6492/packages/dotcom-shared/src/tla-schema/TldrawAppFile.ts#L16
- Kicking guests out of the room if the `shared` flips from `true` to `false`
- Allow guests to not be logged in
- Changing between readonly and readwrite changes the UI and editor mode on guests' machines.
- Simplified routing
  - `/q/local` -> `/q`
  - Remove defunct redirect pages
- Renamed `'temporary'` flag to `'isCreateMode'` and use it to fix the race condition when creating new rooms
- (temporary fix) allow guests to see the document's file name

### Change type

- [ ] `bugfix`
- [ ] `improvement`
- [x] `feature`
- [ ] `api`
- [ ] `other`

### Test plan

1. Create a shape...
2.

- [ ] Unit tests
- [ ] End to end tests

### Release notes

- Fixed a bug with…

### CODE REVIEW

File: apps/dotcom/client/src/tla/providers/TlaProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import { getAssetUrlsByImport } from '@tldraw/assets/imports.vite'
+import { ReactNode } from 'react'
+import { Outlet } from 'react-router-dom'
+import { useValue } from 'tldraw'
+import { AppStateProvider, useApp } from '../hooks/useAppState'
+import { UserProvider } from '../hooks/useUser'
+import '../styles/tla.css'
+import { TlaRootProviders } from './TlaRootProviders'
+
+export const assetUrls = getAssetUrlsByImport()
```
Comment (mimecuvalo) [inline apps/dotcom/client/src/tla/providers/TlaProvider.tsx:11]: looks like this'll need to be merged with the latest stuff btw

File: apps/dotcom/sync-worker/src/TLAppDurableObject.ts [hunk 2]
```diff
@@ -261,4 +281,19 @@ export class TLAppDurableObject {
...
+		userId: string
+	): Promise<'private' | 'view' | 'edit'> {
+		if (!this._userId) {
+			this._userId = userId
+		} else if (userId !== this._userId) {
+			throw new Error('userId mismatch')
+		}
+		const room = await this.getRoom()
+		const file = room.getRecord(fileId) as TldrawAppFile | undefined
+		if (!file?.shared) return 'private'
```
Comment (mimecuvalo) [inline apps/dotcom/sync-worker/src/TLAppDurableObject.ts:296]: later on, when we have more granular sharing controls (i.e. per-user controls) is this going to be able to be refactored easily?

File: apps/dotcom/sync-worker/src/TLDrawDurableObject.ts [hunk 3]
```diff
@@ -513,7 +514,47 @@ export class TLDrawDurableObject {
...
+
+		for (const session of room.getSessions()) {
+			// allow the owner to stay connected
+			if (session.meta.userId === ownerId) continue
+
+			if (!file.shared) {
+				room.closeSession(session.sessionId, TLSyncErrorCloseEventReason.FORBIDDEN)
+			} else if (
+				// if the file is still shared but the readonly state changed, make them reconnect
+				(session.isReadonly && !roomIsReadOnlyForGuests) ||
```
Comment (mimecuvalo) [inline apps/dotcom/sync-worker/src/TLDrawDurableObject.ts:552]: nit: you could do `session.isReadonly !== roomIsReadOnlyForGuests`

**Review by mimecuvalo — APPROVED** [review]
LG 👍 just a question for long term sharing (`file.shared` stuff)

File: apps/dotcom/sync-worker/src/TLAppDurableObject.ts (hunk 2 above)
Comment (ds300) [reply apps/dotcom/sync-worker/src/TLAppDurableObject.ts:296]: Sure

**Pull request merged**